Code_Exercises/9_sycl_of_hell/2_parallel_for.cpp

```diff
@@ -0,0 +1,46 @@
+#include "argparse.hpp"
+#include <sycl/sycl.hpp>
+
+int main(int argc, char **argv) {
+
+  //  _                ___
+  // |_) _. ._ _  _     |  ._  ._     _|_
+  // |  (_| | _> (/_   _|_ | | |_) |_| |_
+  //                           |
+  argparse::ArgumentParser program("2_parallel_for");
+
+  program.add_argument("-g", "--global")
+      .help("Global Range")
+      .default_value(1)
+      .action([](const std::string &value) { return std::stoi(value); });
+
+  try {
+    program.parse_args(argc, argv);
+  } catch (const std::runtime_error &err) {
+    std::cout << err.what() << std::endl;
+    std::cout << program;
+    exit(0);
+  }
+
+  const auto global_range = program.get<int>("-g");
+  //  _                        _
+  // |_) _. ._ _. | |  _  |   |_ _  ._
+  // |  (_| | (_| | | (/_ |   | (_) |
+
+  sycl::queue Q;
+  std::cout << "Running on " << Q.get_device().get_info<sycl::info::device::name>() << std::endl;
+
+  // Similar to 
+  // #pragma OMP PARALLEL FOR
+  // for(int idx=0; idx++; idx < global_range)
+  Q.parallel_for(
+       global_range,
+       [=](sycl::id<1> idx) {
+         // Explicit cast because of printf shenaningan.
+         sycl::ext::oneapi::experimental::printf("Hello, World! World rank %d\n",
```

Comment:
I don't know what you've planned exactly here, but from the outside it doesn't really feel appropriate to me to use implementation-specific extensions in a generic SYCL tutorial.

Reply:
Yeah, I agree... You support printf around you?

If not I need to user the horrible sycl::Steeam but then I need to teach `command group handler` and it always have been a pain...

Reply:
Ah, in the context of moving away from handler I can see the issue.

We don't universally have a variadic printf (and we most certainly will not implement oneAPI interfaces after Intel has set the precedent of implementing AdaptiveCpp extensions but changing the name without technical reason so that the API is incompatible).

But if it's acceptable to just print `"Hello World"`` without rank, I can give you some internal detail function for printing at least strings, which would happen to work for now ;) 
That would be `sycl::detail::print(const char* s)`.

Reply:
Yes, good enough! I can always cast my int into string for the 
```
  Q.parallel_for(global_range, [=](sycl::id<1> idx) {
    // Explicit cast because of printf shenaningan.
    sycl::ext::oneapi::experimental::printf("Hello, World! World rank %d\n", static_cast<int>(idx));
  });
  ``` 
  :) 
  
  Yeah, nested lambda tripped off a lot of people in the previous tutorial... Even more the Fortran dev who saw c++ presently. (but maybe I was just particularly bad at explaining cgh...)

Reply:
Great :) Main reason there's no int supported directly is because I didn't want to spend a lot of time developing a generic device string formatting library :P But just for int it's reasonably simple.

Yeah, I'm all for killing buffer and command group handler ;)
These are just additional unnecessary complexities in my opinion.